src/server/session.ts

```diff
@@ -179,16 +178,13 @@ namespace ts.server {
         public startNew(action: (next: NextStep) => void) {
             this.complete();
             this.requestId = this.operationHost.getCurrentRequestId();
-            this.completed = false;
             this.executeAction(action);
         }
 
         private complete() {
-            if (!this.completed) {
-                if (this.requestId) {
-                    this.operationHost.sendRequestCompletedEvent(this.requestId);
-                }
-                this.completed = true;
+            if (this.requestId) {

```

Comment:
It looks like the server is responsible for sending us these IDs, so it might have happened to always be non-zero when coming from VSCode, but best not to rely on that.